chainer/iterators/multiprocess_iterator.py

```diff
@@ -126,167 +146,285 @@ def serialize(self, serializer):
+        with self._lock:
+            self._status = _Communicator.STATUS_TERMINATE
+            self._batch_queue = []
+            self._not_full_cond.notify()
+            self._reset_count += 1
+
+    # called from thread
+    def check(self):
+        with self._lock:
+            status = self._status
+            self._status = _Communicator.STATUS_CONTINUE
+            prefetch_state = None
+            if status == _Communicator.STATUS_RESET:
+                prefetch_state = self._prefetch_state
+            return status, prefetch_state, self._reset_count
+
+    # called from thread
+    def put(self, batch, prefetch_state, reset_count):
+        with self._lock:
+            if len(self._batch_queue) == self.n_prefetch:
+                self._not_full_cond.wait(_long_time)
+            if reset_count == self._reset_count:
+                self._batch_queue.append((batch, prefetch_state))
+                self._not_empty_cond.notify()
+
+
+class _PrefetchLoop(object):
+
+    def __init__(self, dataset, batch_size, repeat, shuffle,
+                 n_processes, n_prefetch, mem_size, comm,
+                 _interruption_testing):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.repeat = repeat
+        self.shuffle = shuffle
+        self.n_processes = n_processes
+        self.mem_size = mem_size
+        self.comm = comm
+
+        self._allocate_shared_memory()
+        self._pool = None
+
+        self._interruption_testing = _interruption_testing
+
+    def __call__(self):
+
+        if self.mem_bulk is None:
+            if not self._task():
+                return
+
+        self._pool = multiprocessing.Pool(
+            processes=self.n_processes,
+            initializer=_fetch_setup,
+            initargs=(self.dataset, self.mem_size, self.mem_bulk))
+        if self._interruption_testing:
+            pids = self._pool.map(_report_pid, range(self.n_processes))
+            print(' '.join(map(str, pids)))
+            sys.stdout.flush()
+
+        alive = True
+        try:
+            while alive:
+                alive = self._task()
+        finally:
+            self._pool.close()
+            self._pool.join()
+
+    def _task(self):
+        status, prefetch_state, reset_count = self.comm.check()
+        if status == _Communicator.STATUS_RESET:
+            self.prefetch_state = prefetch_state
+        elif status == _Communicator.STATUS_TERMINATE:
+            return False  # stop loop
+
+        indices = self._proceed()
+        if indices is None:  # stop iteration
+            batch = None
+        elif self._pool is None:  # measure mode
+            # To enable terminating the thread on Python 2.7,
+            # we spawn another process.
+            # We don't use Pool here, because creating Pools in short-time
+            # had caused rare deadlock on Python 3.5.2.
+            q = multiprocessing.Queue()
+            p = multiprocessing.Process(target=_measure_run,
```

Comment:
I was afraid that memory leak occurs when the iterator is GC'ed if indexing dataset is very very slow. But such a program doesn't functions at all and the leak cleared when the process terminates, so now I think avoiding this leak doesn't worth the complexity of launching another process.

I will add another commit that doesn't launch a process at measurement step.